Enable parsing TDH_IN_POINTER types

pathtofile opened the pull request:
This merge attempts to enable easy parsing of `TDH_INTYPE_POINTER`, whilst attempting to protect developers from poor decisions.

Whilst `TDH_INTYPE_POINTER` are Pointers, the likelihood the address is accessible to the trace process is extremely low, and at times might even be pointers to kernel memory.
As such I feel any solution needs to make it clear to developers using Krabs that while it is OK to get the pointer for printing or comparison purposes, any attempt at de-referencing it will probably blow up in their face.

If we just enable `.parse<void*>()` it as a `void*`, then a dev could blindly do:
```c
std::uint32_t* pFileobj = (std::uint32_t*)parser.parse<void*>(L"FileObject");

// This will either explode (esp. if a kernel pointer), or point to something else entirely
std::uint32_t fileobj = *pFileobj ;

// If the above didn't explode, then this could be altering something else
fileobj += 5;
```

That's why my attempt at a solution opted to create a new struct type `pointer`, with the reference named `wild_reference`. This makes the developer have to write:
```c
krabs::pointer krabs_fileobj = parser.parse<krabs::pointer>(L"FileObject");

// Have to use 'wild' reference
// Also const prevents alteration
const std::uint32_t* pFileobj_wild = (const std::uint32_t*)(krabs_fileobj.wild_reference);

// This will still go boom, but we gave them a bit a warning.
const std::uint32_t fileobj = *pFileobj_wild;
```

We could also use `dangerous` instead of `wild` if you feel that would be better.

Another option would make `krabs::pointer.wild_reference` an integer instead of a `void*`, make it even more explicit that the pointer isn't for de-referencing.

swannman commented on Microsoft.O365.Security.Native.ETW/EventRecord.hpp:
```diff
@@ -351,6 +351,53 @@ namespace Microsoft { namespace O365 { namespace Security { namespace ETW {
+        }
+
+        /// <summary>
+        /// Attempt to get a Security Identifier (SID) from the specified property name.
```
copy/paste error on the comments?

swannman commented on krabs/krabs/parser.hpp:
```diff
@@ -367,14 +367,14 @@ namespace krabs {
-        static const int sid_start = 16;
+        ULONG sid_start = 16;
+        if (EVENT_HEADER_FLAG_32_BIT_HEADER == (schema_.record_.EventHeader.Flags & EVENT_HEADER_FLAG_32_BIT_HEADER)) {
+            sid_start = 8;
```
thank you for adding this! I wondered whether we needed special handling for 32-bit here.

swannman commented on krabs/krabs/parse_types.hpp:
```diff
@@ -253,6 +253,42 @@ namespace krabs {
+        // Make the reference const and call it wild,
+        // As any attempt to derefernce or alter what it points to
+        // would be very undefined behaviour, as it
+        // is unlikley it is pointing to inside our process' memory
```
typo `unlikely`

swannman commented on krabs/krabs/parse_types.hpp:
```diff
@@ -253,6 +253,42 @@ namespace krabs {
+    */
+    struct pointer {
+        // Make the reference const and call it wild,
+        // As any attempt to derefernce or alter what it points to
```
typo `dereference`

swannman commented:
cc @kylereedmsft who may have thoughts on `krabs::pointer.wild_reference`. The current proposal looks reasonable to me but I'm not sure whether there is already an idiomatic way to handle this?

kallanreed commented on krabs/krabs/parse_types.hpp:
```diff
@@ -253,6 +253,42 @@ namespace krabs {
+        // As any attempt to derefernce or alter what it points to
+        // would be very undefined behaviour, as it
+        // is unlikley it is pointing to inside our process' memory
+        const void* wild_reference;
```
Kind of a strange name. Few thoughts: 1) why even store it as an actual pointer at all? It's probably never a good idea to use it as a pointer in code. You could just keep a uint64_t value and let the top bytes be empty on x86. Just call it value or something. Alternately, you could make wrap this in a a type that doesn't allow dereferencing at all (just overload operator* and make it assert, but ultimately you'll still want to see the value at some point, so it seems like using an integer may be better

If you really want a pointer (so things like pointer printf specifiers work), I'd still suggest calling it `value` instead of `wild_reference`.

jdu2600 replied:
Hey Pat,

Nice work.  :-)

I think that using a uint64_t should be enough of a warning to the developer.
And perhaps just 'address' if it needs a name?

John

For reference -
perfview's TraceEvent defines an 'Address' type which is just a UInt64.
And the default string representation is 0x%x.

https://github.com/microsoft/perfview/blob/9d8a9b90c5ed917da1b9bdea98210114a270e633/src/TraceEvent/TraceEvent.cs#L20
https://github.com/microsoft/perfview/blob/9d8a9b90c5ed917da1b9bdea98210114a270e633/src/TraceEvent/TraceEvent.cs#L978

kallanreed commented on krabs/krabs/parse_types.hpp:
```diff
@@ -253,6 +253,42 @@ namespace krabs {
+            // If 32-Bit, first parse as a uint32
+            // Then we can 'cast' that as pointer
+            if (size_in_bytes == sizeof(uint32_t)) {
+                uint64_t tmp = *(uint32_t*)bytes;
```
I think we prefer static_casts in this code. Just be consistent with the rest of the file

pathtofile commented:
@jdu2600 @kallanreed @swannman - So I'm happy with instead making it just an `uint64_t`, as I also don't really think people should be attempting to use the pointer. Summary of changes:
* Changed `struct pointer` to parse and hold address as a `uint64_t` instead of a `void*`
* Renamed address variable to "address", to make the name clearer.
* Fixed various typos from bad copy/paste or keyboarding.

pathtofile commented:
RE: https://github.com/microsoft/krabsetw/pull/132#discussion_r423197793 @kallanreed it had to be a `reintepret_cast` as it doesn't like the `byte*`->`uint32_t*` conversion, but I took your comment to mean 'be consistent with using c++ style casts' and fixed it

swannman commented:
@kallanreed any review comments before I merge?

kylereedmsft commented:
Looks ok to me.



________________________________
From: Matt Swann <notifications@github.com>
Sent: Monday, May 18, 2020 6:20:51 PM
To: microsoft/krabsetw <krabsetw@noreply.github.com>
Cc: Kyle Reed <kylereed@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [microsoft/krabsetw] Enable parsing TDH_IN_POINTER types (#132)


@kallanreed<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fkallanreed&data=02%7C01%7Ckylereed%40microsoft.com%7C3a8e947c5418450e062f08d7fb92dee0%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637254480570642630&sdata=fWwQgoClrl2CzvIFniHbrihOHt2xAO2Fc11BiCAKfvY%3D&reserved=0> any review comments before I merge?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2Fkrabsetw%2Fpull%2F132%23issuecomment-630519308&data=02%7C01%7Ckylereed%40microsoft.com%7C3a8e947c5418450e062f08d7fb92dee0%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637254480570652629&sdata=%2FwBfNtG091UHe3q1l%2BhdXLJj%2F2XyjEumuV7qFIPMFp4%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAFBGWRDVEBG3APQL6PN6UWDRSHNHFANCNFSM4M5G7S4A&data=02%7C01%7Ckylereed%40microsoft.com%7C3a8e947c5418450e062f08d7fb92dee0%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637254480570662628&sdata=4toD%2Bdol1ryfE3X84ZebdOOgCZdOPx3bgM0aqkJRKF4%3D&reserved=0>.

pathtofile commented:
@swannman updated the nuspec
